File: LivrusVirtus/wxWindows/wx.NET/VCPP.NET/RotinasGlobais.cpp
#include "StdAfx.h"
#include "RotinasGlobais.h"
#include "Principal.h"
#include "Resource.h"

namespace Livrus
{
  RotinasGlobais::RotinasGlobais(void){
    gConexao = new System::Data::OleDb::OleDbConnection();
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ValidaFloat(String* sNum)
  {
   if (sNum->Equals(S"")==false) {
  	try {
  	  Single::Parse(sNum);
  	  return true;
  	} catch(Exception* e) {
      wx::MessageDialog::MessageBox(Resource::STR_NUMINV, "Erro", 
        wx::Dialog::wxOK | wx::Dialog::wxICON_ERROR );
  	  return false;
  	}
   } else
     return false;
  } 
  //---------------------------------------------------------------------------
  String* RotinasGlobais::VirgulaParaPonto(String* sNum, Boolean revertido)
  {
   if (revertido)
     return sNum->Replace(S".",S",");
   else
     return sNum->Replace(S",",S".");
  }
  //---------------------------------------------------------------------------
  int RotinasGlobais::modulo11(String* svalor, int ndig)
  {
    int ncont, nsoma, nvalor, npos;
    String *slocal, *sres;
    
     ncont=0;
     slocal="";
     while (ncont < 11) {
       slocal=String::Concat(svalor->Substring(ncont,1), slocal);
       ncont++;
     }
     nsoma=0;
     if (ndig == 1)
       ncont=2;
     else
       ncont=1;
     npos=2;
     while (ncont < 11) {
       sres = slocal->Substring(ncont,1);
       nvalor = Int32::Parse(sres);
       nsoma=nsoma + (nvalor * npos);
       ncont++;
       npos++;
     }
     nsoma=(11 - (nsoma % 11));
    
     if (nsoma > 9)
       return 0;
     else
       return nsoma;
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ValidaCPF(String* cpf){
    String* scpf;
  
    if (cpf->Length==14) {
      scpf=String::Concat(cpf->Substring(0,3),cpf->Substring(4,3),
        cpf->Substring(8,3),cpf->Substring(12,2));
      if ( (modulo11(scpf,1)==Int32::Parse(cpf->Substring(12,1))) &&
      (modulo11(scpf,2)==Int32::Parse(cpf->Substring(13,1))) )
      	return true;
      else
      	return false;
    } else
      return false;
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ValidaISBN(String* isbn)
  {
    String *sisbn, *sdigito, *snum;
    int nvalor, nsoma, ncont;
    
    if (isbn->Length==13) {
      sdigito = isbn->Substring(isbn->Length-1, 1);
      sisbn="";
      for(ncont = 0; ncont < isbn->Length-1; ncont++)  {
        if (isbn->Substring(ncont,1)->Equals(S"-")==false)
          sisbn = String::Concat(sisbn, isbn->Substring(ncont,1));
      }
      if (sdigito->Equals(S"X"))
        nsoma = 10;
      else {
        nvalor = Int32::Parse(sdigito);
        nsoma = nvalor;
      }
      for(ncont = 1; ncont <= 9; ncont++)  {
        snum = sisbn->Substring(ncont-1, 1);
        nvalor = Int32::Parse(snum);
        nsoma=nsoma + (nvalor * (11 - ncont));
      }
      if ((nsoma % 11) == 0)
        return true;
      else
        return false;
    } else
      return false;
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::validaDataHora(String* sDataHora, Boolean bSoData) {
    Boolean bvalida, bdatavalida, bhoravalida;
    String *sdia, *smes, *sano, *shora, *smin, *sseg;
    int ndia, nmes, nano, nhora, nmin, nseg, nudiames, ntam;
  
    bvalida = false;
    bdatavalida = false;
    bhoravalida = false;
    nudiames = 0;
    if (bSoData==true) ntam=10;
    else ntam=19;

    if (sDataHora->Length==ntam) {
  	sdia = sDataHora->Substring(0,2);
  	ndia = Int32::Parse(sdia);
  	smes = sDataHora->Substring(3,2);
  	nmes = Int32::Parse(smes);
  	sano = sDataHora->Substring(6,4);
  	nano = Int32::Parse(sano);
  
  	if ((nano >= 1) && (nano <= 9999)) {
  	  if ((nmes >= 1) && (nmes <= 31)) {
  		  switch(nmes) {
  		   case 1: case 3: case 5: case 7: case 8:
  		   case 10: case 12: nudiames = 31; break;
  		   case 2: {
  			   /* ano bissexto */
                 if ((nano % 4)==0) nudiames = 29;
                 else nudiames = 28;
  			  } break;
             case 4: case 6: case 9: case 11: nudiames = 30; break;
            }
  		  if ((ndia >=1) && (ndia <=nudiames))
               bdatavalida = true;
        }
      }
  
      if (bSoData) {
  	   if (bdatavalida)
  		   bvalida = true;
       else
  	     bvalida = false;
  	} else {
  	  if (bdatavalida) {
  		  shora = sDataHora->Substring(11,2);
  		  nhora = Int32::Parse(shora);
  		  smin = sDataHora->Substring(14,2);
  		  nmin = Int32::Parse(smin);
  		  sseg = sDataHora->Substring(17,2);
  		  nseg = Int32::Parse(sseg);
  
  		  if ((nhora >= 0) && (nhora <= 23)) {
  			if ((nmin >= 0) && (nmin <= 59)) {
  			  if ((nseg >= 0) && (nseg <= 59))
  				 bhoravalida = true;
  			}
  		  }
  
  		  if (bhoravalida)
  		   bvalida = true;
  		  else
  		   bvalida = false;
  	  } else
          bvalida = false;
      }
    } else
      bvalida = false;

    return bvalida;
  }
  //---------------------------------------------------------------------------
  String* RotinasGlobais::retDataHoraStr(Boolean bData, Boolean bHora) {
    String* sDataHora = "";
    
    if (bData && bHora)
      sDataHora = DateTime::Now.ToString("dd/MM/yyyy hh:mm:ss");
    else if (bData)
      sDataHora = DateTime::Now.ToString("dd/MM/yyyy");
    else if (bHora)
      sDataHora = DateTime::Now.ToString("hh:mm:ss");
      
    return sDataHora;
  }
  //---------------------------------------------------------------------------
  String* RotinasGlobais::ConverteDataHoraStr(String* sData, String* formato) {
  	System::DateTime dtNasc = System::DateTime::Parse(sData);
  	return dtNasc.ToString(formato);
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ConectaBanco(String* sFonteDeDados){
		gConexao->ConnectionString = sFonteDeDados;
    gConexao->Open();

    if (gConexao->get_State() == ConnectionState::Open)
      return true;
    else {
      MessageDialog::MessageBox(Resource::STR_NCONBANCO, "Erro", 
        Dialog::wxOK|Dialog::wxICON_ERROR );
      return false;
    }
  }
  //---------------------------------------------------------------------------    
  Void RotinasGlobais::DisconectaBanco(){
    gConexao->Close();
  }
  //---------------------------------------------------------------------------    
  Boolean RotinasGlobais::AtualizaDados(OleDbCommand* Arquiva, String* sTextoSql){
	  Arquiva = gConexao->CreateCommand();
	  Arquiva->CommandText = sTextoSql;
		if (Arquiva->ExecuteNonQuery() > 0) {
		  Arquiva->Dispose();
      return true;
    } else {
      //MessageDialog::MessageBox(Resource::STR_NCEXECSQL, "Erro", 
      //  Dialog::wxOK|Dialog::wxICON_ERROR );
      return false;
    }
  }
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ConsultaDados(OleDbCommand* Consulta, String* sTextoSql)
  {
  Boolean nTemLinhas = false;
  
    Consulta = gConexao->CreateCommand();
    Consulta->CommandText = sTextoSql;
  
    OleDbDataReader* Leitor = Consulta->ExecuteReader();
    nTemLinhas = Leitor->HasRows;
    Leitor->Close();
    return nTemLinhas;  
  } 
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ConsultaDadosLista(ArrayList* Registros, String* sTextoSql)
  {
  	Boolean nTemLinhas = false;
  
    OleDbCommand* Consulta = gConexao->CreateCommand();
    Consulta->CommandText = sTextoSql;
  
    Registros->Clear();
    OleDbDataReader* Leitor = Consulta->ExecuteReader();
    while (Leitor->Read()) {
  	 ArrayList* Campos = new ArrayList();
  	 for(int cont=0; cont < Leitor->FieldCount; cont++) {
  	   Campos->Add(Leitor->GetValue(cont)->ToString());
  	 }
  	 Registros->Add(Campos);
    }
    nTemLinhas = Leitor->HasRows;
    Leitor->Close();
    return nTemLinhas;
  
  }  
  //---------------------------------------------------------------------------
  Boolean RotinasGlobais::ValidaCampo(String* sCampo,String* sMensagemErro)
  {
    if (sCampo->Equals(S"")) {
      MessageDialog::MessageBox(sMensagemErro, S"Erro", 
        Dialog::wxOK|Dialog::wxICON_ERROR );
    	return true;
    } else
    	return false;
  }
  //---------------------------------------------------------------------------
  String* RotinasGlobais::ConsultaCampoDesc(ArrayList* Campos,
  				String* sTextoSql, String* sMensagemErro)
  {  
    OleDbCommand* Consulta = gConexao->CreateCommand();
    Consulta->CommandText = sTextoSql;
  
    Campos->Clear();
    OleDbDataReader* Leitor = Consulta->ExecuteReader();
    if (Leitor->Read()) {
  	 for(int cont=0; cont < Leitor->FieldCount; cont++) {
  	   Campos->Add(Leitor->GetValue(cont)->ToString());
  	 }
    }
    bool nTemLinhas = Leitor->HasRows;
    Leitor->Close();
  
    if (nTemLinhas)
      return Campos->Item[1]->ToString();
    else {
      MessageDialog::MessageBox(sMensagemErro, S"Erro", 
        Dialog::wxOK|Dialog::wxICON_ERROR );
    	return "";
    }
  } 
  //---------------------------------------------------------------------------
  int RotinasGlobais::PesquisaItemLista(wx::ListBox* lstLista, String* sItem){
    int bret = -1;
    int nLinha;
    for (nLinha = 0; nLinha < lstLista->Count; nLinha++){
      if (lstLista->GetString(nLinha) == sItem) {
        bret = 0;
      }        
    }
    return bret;
  }
  //---------------------------------------------------------------------------
  Void RotinasGlobais::AdicionaItemLista(wx::ListBox* lstLista,
  	 ArrayList* slCodLista, wx::TextCtrl* sCampoCodigo, wx::TextCtrl* sCampoDesc)
  {
    String* sItems;
  
    if (sCampoDesc->Value->Equals(S"")==false) {
      sItems=String::Concat(sCampoCodigo->Value,S" - ",sCampoDesc->Value);
      if (PesquisaItemLista(lstLista, sItems)==-1) {
        slCodLista->Add(sCampoCodigo->Value);
        lstLista->Append(sItems);
      }
      sCampoCodigo->Clear();
      sCampoDesc->Clear();
    }
  }
  //---------------------------------------------------------------------------
  Void RotinasGlobais::RemoveItemLista(wx::ListBox* lstLista, ArrayList* slCodLista)
  {
    int nSelIndice = 0;
    if (lstLista->Count > 0){
      nSelIndice = lstLista->Selection;
      if (lstLista->Selected(nSelIndice)) {
        slCodLista->RemoveAt(nSelIndice);
        lstLista->Delete(nSelIndice);
      }
    }
  }
  //---------------------------------------------------------------------------
   Void RotinasGlobais::AdicionaValoresLista(String* sTextoSql,
     wx::ListBox* lstLista, ArrayList* slCodLista){
    lstLista->Clear();
    slCodLista->Clear();
  
    OleDbCommand* Consulta = gConexao->CreateCommand();
    Consulta->CommandText = sTextoSql;
  
    OleDbDataReader* Leitor = Consulta->ExecuteReader();
    while (Leitor->Read()) {
  	  slCodLista->Add(Leitor->GetValue(0)->ToString());
  	  lstLista->Append(String::Concat(Leitor->GetValue(0)->ToString()," - ",
  	  Leitor->GetValue(1)->ToString()));
    }
    Leitor->Close();
  
  }
  //---------------------------------------------------------------------------
  Void RotinasGlobais::AdicionaColunasGrid(wx::Grid* grid, int nPosIni, 
    int nQtdCols, int nLargura __gc[], String* sRotulos __gc[]){
    int nPos, nCont;
    
    if (nPosIni > 0){
      grid->AppendCols(nQtdCols);
    }
    nCont = 0;
    for(nPos = nPosIni; nPos < (nPosIni+nQtdCols); nPos++){
      grid->SetColSize(nPos, nLargura[nCont]);
      grid->SetColLabelValue(nPos, sRotulos[nCont]);
      nCont = nCont + 1;
    }
  }
  //---------------------------------------------------------------------------  
  Void RotinasGlobais::AdicionaLinhasGrid(wx::Grid* grid, ArrayList* lstLinhas, 
    int nQtdCol){
    int nCol, nLinha;
    ArrayList *lstCampos;

    grid->DeleteRows(0, grid->NumberRows, true);
    grid->AppendRows(lstLinhas->Count, true);
    for (nLinha = 0; nLinha < lstLinhas->Count; nLinha++){
		  lstCampos = dynamic_cast<ArrayList*>(lstLinhas->Item[nLinha]);
      for (nCol = 0; nCol < nQtdCol; nCol++){
        grid->SetCellValue(nLinha, nCol, lstCampos->Item[nCol]->ToString());
      }
    }
  }
  //---------------------------------------------------------------------------
}

File: LivrusVirtus/wxWindows/wx.NET/Delphi.NET/fontes/RotinasGlobais.pas
unit RotinasGlobais;

interface

uses System.Collections, System.Windows.Forms, System.Data, 
  System.Data.OleDb, wx;

type
  TRotinasGlobais = class
  private
    { Private Declarations }
  public
    sUsuarioLogin, sUsuarioSenha: string;
    sCodigoSelecionado, sPreco, sQtdEstoque, sBarraLogin: string;
    gConexao: System.Data.OleDb.OleDbConnection;
    constructor Create;
    function ValidaFloat(sNum: string): boolean;
    function ValidaCampo(sCampo, sMensagemErro: string): boolean;
    function VirgulaParaPonto(sNum: string; revertido: boolean): string;
    function modulo11(svalor: string; ndig: integer): integer;
    function ValidaCPF(cpf: string): boolean;
    function ValidaISBN(isbn: string): boolean;
    function validaDataHora(sDataHora: string; bSoData: boolean): boolean;
    function retDataHoraStr(bData, bHora: boolean): string;
    function ConverteDataHoraStr(sData, formato: String): String;
    function ConectaBanco(sFonteDeDados: string): boolean;
    procedure DisconectaBanco();
    function AtualizaDados(Arquiva: OleDbCommand; sTextoSql: string): boolean;
    function ConsultaDados(Consulta: OleDbCommand; sTextoSql: string): boolean;
    function ConsultaDadosLista(Registros: ArrayList; sTextoSql: string): boolean;
    function ConsultaCampoDesc(Campos: ArrayList; sTextoSql,
      sMensagemErro:String): String;
    function PesquisaItemLista(lstLista: wx.ListBox; sItem: string): integer;
    procedure AdicionaItemLista(lstLista: wx.ListBox; slCodLista: ArrayList;
      sCampoCodigo: wx.TextCtrl; sCampoDesc: wx.TextCtrl);
    procedure RemoveItemLista(lstLista: wx.ListBox; slCodLista: ArrayList);
    procedure AdicionaValoresLista(sTextoSql: String; lstLista: wx.ListBox;
      slCodLista: ArrayList);
    procedure AdicionaColunasGrid(grid: wx.Grid; nPosIni, 
      nQtdCols: integer; nLargura: array of integer; sRotulos: array of string);   
    procedure AdicionaLinhasGrid(grid: wx.Grid; 
      lstLinhas: ArrayList; nQtdCol: integer);
  end;

var
  Rotinas: TRotinasGlobais;

const
  sepDTHR: string = '#'; //SOMENTE ACCESS 
  //sepDTHR: string = ''''; //OUTROS BANCOS 

implementation

uses Principal, Resource;

constructor TRotinasGlobais.Create;
begin
  inherited Create;
  gConexao := System.Data.OleDb.OleDbConnection.Create();
end;

function TRotinasGlobais.ValidaFloat(sNum: string): boolean;
begin
 if sNum<>'' then begin
    try
      Convert.ToSingle(sNum);
      result:=true;
    except
      MessageDialog.MessageBox(Res.STR_NUMINV, 'Erro', 
        Dialog.wxOK or Dialog.wxICON_ERROR );
      result:=false;
    end;
 end else
   result:=false;
end;

function TRotinasGlobais.VirgulaParaPonto(sNum: string;
revertido: boolean): string;
begin
 if revertido then
   result:= sNum.Replace('.',',')
 else
   result:= sNum.Replace(',', '.');
end;

function TRotinasGlobais.modulo11(svalor: string; ndig: integer): integer;
var ncont, nsoma, nvalor, npos: integer;
    sres, slocal: string;
begin
 ncont:=1;
 while ncont <= 11 do begin
   slocal:=copy(svalor,ncont,1) + slocal;
   ncont:=ncont+1;
 end;
 nsoma:=0;
 if ndig=1 then
   ncont:=3
 else
   ncont:=2;
 npos:=2;
 while ncont <= 11 do begin
   sres:=copy(slocal,ncont,1);
   nvalor:=Int32.Parse(sres);
   nsoma:=nsoma + (nvalor * npos);
   ncont:=ncont+1;
   npos:=npos+1;
 end;
 nsoma:=(11 - (nsoma mod 11));

if nsoma > 9 then
  modulo11:=0
else
  modulo11:=nsoma;
end;

function TRotinasGlobais.ValidaCPF(cpf: string): boolean;
var scpf: string;
begin

if (length(cpf)=14) then begin
  scpf:=copy(cpf,1,3)+copy(cpf,5,3)+copy(cpf,9,3)+copy(cpf,13,2);
  if ( (modulo11(scpf,1)=Int32.parse(copy(cpf,13,1))) and
  (modulo11(scpf,2)=Int32.parse(copy(cpf,14,1))) ) then
    result:=true
  else
    result:=false;
end else
  result:=false;
end;

function TRotinasGlobais.ValidaISBN(isbn: string): boolean;
var snum, sdigito, sisbn: string;
 nvalor, nsoma, ncont: integer;
begin
if (length(isbn)=13) then begin
  sdigito := copy(isbn, length(isbn), 1);
  sisbn:='';
  for ncont:=1 to length(isbn)-1 do begin
    if (copy(isbn,ncont,1)<>'-') then
      sisbn:=sisbn+copy(isbn,ncont,1);
  end;
  if sdigito = 'X' then
    nsoma := 10
  else begin
    nvalor:=Int32.Parse(sdigito);
    nsoma := nvalor;
  end;
  for ncont:=1 to 9 do begin
    snum:=copy(sisbn, ncont, 1);
    nvalor:=Int32.Parse(snum);
    nsoma:=nsoma + (nvalor * (11 - ncont));
  end;
  If (nsoma mod 11) = 0 Then
    validaISBN:=true
  else
    validaISBN:=false;
end else
  validaISBN:=false;
end;

function TRotinasGlobais.validaDataHora(sDataHora: string;
  bSoData: boolean): boolean;
var bvalida, bdatavalida, bhoravalida: boolean;
    sdia, smes, sano, shora, smin, sseg: string;
    ndia, nmes, nano, nhora, nmin, nseg, nudiames: integer;
    nStatus, ntam: integer;
begin
bdatavalida:=false;
bhoravalida:=false;
nudiames:=0;

if bSoData=true then ntam:=10
else ntam:=19;

if length(sDataHora)=ntam then begin
  sdia:=copy(sDataHora,1,2);
  Val(sdia,ndia,nStatus);
  smes:=copy(sDataHora,4,2);
  Val(smes,nmes,nStatus);
  sano:=copy(sDataHora,7,4);
  Val(sano,nano,nStatus);

  if (nano >= 1) and (nano <= 9999) then begin
    if (nmes >= 1) and (nmes <= 31) then begin
        case nmes of
         1, 3, 5, 7, 8, 10, 12: nudiames:=31;
         2: begin
             { ano bissexto }
             if (nano mod 4)=0 then nudiames:=29
             else nudiames:=28;
            end;
         4, 6, 9, 11: nudiames:=30;
        end;
        if (ndia >=1) and (ndia <=nudiames) then
           bdatavalida:=true;
    end;
  end;

  if bSoData then begin
     if bdatavalida then bvalida:=true
     else bvalida:=false;
  end else begin
    if bdatavalida then begin
        shora:=copy(sDataHora,12,2);
        Val(shora,nhora,nStatus);
        smin:=copy(sDataHora,15,2);
        Val(smin,nmin,nStatus);
        sseg:=copy(sDataHora,18,2);
        Val(sseg,nseg,nStatus);

        if (nhora >= 0) and (nhora <= 23) then begin
          if (nmin >= 0) and (nmin <= 59) then begin
            if (nseg >= 0) and (nseg <= 59) then
               bhoravalida:=true;
          end;
        end;

        if bhoravalida then bvalida:=true
        else bvalida:=false;
    end else
      bvalida:=false;
  end;
end else 
  bvalida:=false;

result:=bvalida;
end;

function TRotinasGlobais.retDataHoraStr(bData, bHora: boolean): string;
var sDataHora: string;
  sTempo: DateTime;
begin
  sDataHora:='';
  sTempo := DateTime.Now;
  
  if (bData) and (bHora) then
    sDataHora := sTempo.ToString('dd/MM/yyyy hh:mm:ss')
  else if bData then
    sDataHora := sTempo.ToString('dd/MM/yyyy')
  else if bHora then
    sDataHora := sTempo.ToString('hh:mm:ss');
  
  result:=sDataHora;
end;

function TRotinasGlobais.ConverteDataHoraStr(sData, formato: String): String;
var dtNasc: DateTime;
begin
  dtNasc:=DateTime.Parse(sData);
  result:=dtNasc.ToString(formato);
end;

function TRotinasGlobais.ConectaBanco(sFonteDeDados: string): boolean;
begin
	gConexao.ConnectionString := sFonteDeDados;
  gConexao.Open();

  if gConexao.State = ConnectionState.Open then
    result:=true
  else begin
    MessageDialog.MessageBox(Res.STR_NCONBANCO, 'Erro', 
      Dialog.wxOK or Dialog.wxICON_ERROR );
    result:=false;
  end;
end;

procedure TRotinasGlobais.DisconectaBanco();
begin
  gConexao.Close();
end;

function TRotinasGlobais.AtualizaDados(Arquiva: OleDbCommand; 
  sTextoSql: string): boolean;
begin
  Arquiva := gConexao.CreateCommand();
  Arquiva.CommandText := sTextoSql;
	if Arquiva.ExecuteNonQuery() > 0 then begin
	  Arquiva.Dispose();
    result:=true;
  end else begin
    //MessageDialog.MessageBox(Res.STR_NCEXECSQL, 'Erro', 
    //  Dialog.wxOK or Dialog.wxICON_ERROR );
    result:=false;
  end;
end;

function TRotinasGlobais.ConsultaDados(Consulta: OleDbCommand;
sTextoSql: string): boolean;
var nTemLinhas: boolean;
  Leitor: OleDbDataReader;
begin
  Consulta := gConexao.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Leitor := Consulta.ExecuteReader;
  nTemLinhas := Leitor.HasRows;
  Leitor.Close;
  ConsultaDados := nTemLinhas;
end;

function TRotinasGlobais.ConsultaDadosLista(Registros: ArrayList;
  sTextoSql: string): boolean;
var nTemLinhas: boolean;
  Consulta: OleDbCommand;
  Leitor: OleDbDataReader;
  Campos: ArrayList;
  cont: integer;
begin
  Consulta := gConexao.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Registros.Clear;
  Leitor := Consulta.ExecuteReader;
  while Leitor.Read do begin
   Campos := ArrayList.Create;
   for cont:=0 to (Leitor.FieldCount - 1) do begin
     Campos.Add(Leitor.GetValue(cont).ToString);
   end;
   Registros.Add(Campos);
  end;
  nTemLinhas := Leitor.HasRows;
  Leitor.Close;
  ConsultaDadosLista := nTemLinhas;
end;

function TRotinasGlobais.ValidaCampo(sCampo, sMensagemErro: string): boolean;
begin
result:=false;
  if sCampo='' then begin
    MessageDialog.MessageBox(sMensagemErro, 'Erro', 
      Dialog.wxOK or Dialog.wxICON_ERROR );
    result:=true;
  end;
end;

function TRotinasGlobais.ConsultaCampoDesc(Campos: ArrayList;
sTextoSql, sMensagemErro:String): String;
var Consulta: OleDbCommand;
  Leitor: OleDbDataReader;
  cont: integer;
  nTemLinhas: boolean;
begin
  Consulta:= gConexao.CreateCommand;
  Consulta.CommandText:= sTextoSql;

  Campos.Clear;
  Leitor:=Consulta.ExecuteReader;
  if Leitor.Read then begin
	 for cont:= 0 to Leitor.FieldCount - 1 do begin
	   Campos.Add(Leitor.GetValue(cont).ToString);
	 end;
  end;
  nTemLinhas:=Leitor.HasRows;
  Leitor.Close;

  if nTemLinhas then
	 result:=Campos.Item[1].ToString
  else begin
    MessageDialog.MessageBox(sMensagemErro, 'Erro', 
      Dialog.wxOK or Dialog.wxICON_ERROR );
	  result:='';
  end;
end;

function TRotinasGlobais.PesquisaItemLista(lstLista: wx.ListBox; 
  sItem: string): integer;
var bret, nLinha: integer;
begin
  bret := -1;
  for nLinha := 0 to lstLista.Count do begin
    if lstLista.GetString(nLinha) = sItem then begin
      bret := 0;
    end;        
  end;
  result:=bret;
end;

procedure TRotinasGlobais.AdicionaItemLista(lstLista: wx.ListBox;
slCodLista: ArrayList; sCampoCodigo: wx.TextCtrl; sCampoDesc: wx.TextCtrl);
var sItems: String;
begin
  if sCampoDesc.Value<>'' then begin
     sItems:=sCampoCodigo.Value + ' - ' + sCampoDesc.Value;
     if PesquisaItemLista(lstLista, sItems)=-1 then begin
    	slCodLista.Add(sCampoCodigo.Value);
    	lstLista.Append(sItems);
     end;
     sCampoCodigo.Clear;
     sCampoDesc.Clear;
  end;
end;

procedure TRotinasGlobais.RemoveItemLista(lstLista: wx.ListBox;
  slCodLista: ArrayList);
var nSelIndice: integer;
begin
  if lstLista.Count > 0 then begin
    nSelIndice := lstLista.Selection;
    if lstLista.Selected(nSelIndice) then begin
      slCodLista.RemoveAt(nSelIndice);
      lstLista.Delete(nSelIndice);
    end;
  end;
end;

procedure TRotinasGlobais.AdicionaValoresLista(sTextoSql: String;
  lstLista: wx.ListBox; slCodLista: ArrayList);
var Consulta: OleDbCommand;
Leitor: OleDbDataReader;
begin
 lstLista.Clear;
 slCodLista.Clear;

  Consulta:= gConexao.CreateCommand;
  Consulta.CommandText := sTextoSql;

  Leitor:= Consulta.ExecuteReader;
  while Leitor.Read do begin
    slCodLista.Add(Leitor.GetValue(0).ToString);
    lstLista.Append(Leitor.GetValue(0).ToString+' - '+
    Leitor.GetValue(1).ToString);
  end;
  Leitor.Close;
end;

procedure TRotinasGlobais.AdicionaColunasGrid(grid: wx.Grid; nPosIni, 
  nQtdCols: integer; nLargura: array of integer; sRotulos: array of string);
var nPos, nCont: integer;
begin    
  if nPosIni > 0 then begin
    grid.AppendCols(nQtdCols);
  end;
  nCont := 0;
  for nPos := nPosIni to (nPosIni+nQtdCols-1) do begin
    grid.SetColSize(nPos, nLargura[nCont]);
    grid.SetColLabelValue(nPos, sRotulos[nCont]);
    nCont := nCont + 1;
  end;
end;

procedure TRotinasGlobais.AdicionaLinhasGrid(grid: wx.Grid; 
  lstLinhas: ArrayList; nQtdCol: integer);
var nCol, nLinha: integer;
   lstCampos: ArrayList;
begin
  grid.DeleteRows(0, grid.NumberRows, true);
  grid.AppendRows(lstLinhas.Count, true);
  for nLinha := 0 to lstLinhas.Count-1 do begin
	  lstCampos := ArrayList(lstLinhas[nLinha]);
    for nCol:= 0 to nQtdCol-1 do begin
      grid.SetCellValue(nLinha, nCol, lstCampos[nCol].ToString());
    end;
  end;
end;

end.
